File: go/ingest/reconcile_test.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"context"
	"database/sql"
	"fmt"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	"github.com/jeffs-brain/memory/go/brain"
	"github.com/jeffs-brain/memory/go/search"
	"github.com/jeffs-brain/memory/go/store/mem"

	_ "modernc.org/sqlite"
)

// testSetup creates a fresh in-memory store, search index, and
// reconciler for a single test case. The reindexFn records which
// paths were re-indexed instead of performing real work.
type testSetup struct {
	store       *mem.Store
	index       *search.Index
	reconciler  *Reconciler
	reindexed   []brain.Path
	reindexedMu sync.Mutex
}

func newTestSetup(t *testing.T) *testSetup {
	t.Helper()
	store := mem.New()
	t.Cleanup(func() { _ = store.Close() })

	db, err := sql.Open("sqlite", ":memory:")
	if err != nil {
		t.Fatalf("opening sqlite: %v", err)
	}
	t.Cleanup(func() { _ = db.Close() })

	idx, err := search.NewIndex(db, store)
	if err != nil {
		t.Fatalf("creating search index: %v", err)
	}

	ts := &testSetup{
		store: store,
		index: idx,
	}

	rec, err := NewReconciler(ReconcileConfig{
		Store:      store,
		Index:      idx,
		MaxRepairs: 1000,
		ReindexFn: func(ctx context.Context, p brain.Path) error {
			ts.reindexedMu.Lock()
			ts.reindexed = append(ts.reindexed, p)
			ts.reindexedMu.Unlock()
			// Simulate re-indexing by calling Index.Update which scans the store.
			return idx.Update(ctx)
		},
	})
	if err != nil {
		t.Fatalf("creating reconciler: %v", err)
	}
	ts.reconciler = rec
	return ts
}

// writeStoreDoc writes a markdown document to the store at the
// canonical raw/documents path, simulating a successful ingest.
func (ts *testSetup) writeStoreDoc(t *testing.T, slug string, body string) brain.Path {
	t.Helper()
	p := brain.RawDocument(slug)
	content := "---\ntitle: \"" + slug + "\"\n---\n\n" + body + "\n"
	if err := ts.store.Write(context.Background(), p, []byte(content)); err != nil {
		t.Fatalf("writing store doc %s: %v", slug, err)
	}
	return p
}

// indexDoc writes a document to the store AND indexes it via Update().
func (ts *testSetup) indexDoc(t *testing.T, slug string, body string) brain.Path {
	t.Helper()
	p := ts.writeStoreDoc(t, slug, body)
	if err := ts.index.Update(context.Background()); err != nil {
		t.Fatalf("indexing %s: %v", slug, err)
	}
	return p
}

func TestReconcile_detects_missing_from_index(t *testing.T) {
	ts := newTestSetup(t)

	// Write a document to the store but do NOT index it.
	ts.writeStoreDoc(t, "missing-doc", "This document was stored but never indexed.")

	report, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("RunOnce: %v", err)
	}

	if report.TotalDocuments != 1 {
		t.Errorf("expected 1 total document, got %d", report.TotalDocuments)
	}
	if !report.DriftDetected {
		t.Errorf("expected drift to be detected")
	}
	if report.MissingReindexed != 1 {
		t.Errorf("expected 1 missing reindexed, got %d", report.MissingReindexed)
	}
	if report.OrphanedDeleted != 0 {
		t.Errorf("expected 0 orphans deleted, got %d", report.OrphanedDeleted)
	}
	ts.reindexedMu.Lock()
	if len(ts.reindexed) != 1 {
		t.Errorf("expected 1 reindex call, got %d", len(ts.reindexed))
	}
	ts.reindexedMu.Unlock()
}

func TestReconcile_detects_orphaned_in_index(t *testing.T) {
	ts := newTestSetup(t)

	// Index a document, then delete it from the store. The index entry
	// becomes orphaned.
	p := ts.indexDoc(t, "orphan-doc", "Content that will be orphaned.")
	if err := ts.store.Delete(context.Background(), p); err != nil {
		t.Fatalf("deleting store doc: %v", err)
	}

	report, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("RunOnce: %v", err)
	}

	if !report.DriftDetected {
		t.Errorf("expected drift to be detected")
	}
	if report.OrphanedDeleted < 1 {
		t.Errorf("expected at least 1 orphan deleted, got %d", report.OrphanedDeleted)
	}
	if report.MissingReindexed != 0 {
		t.Errorf("expected 0 missing reindexed, got %d", report.MissingReindexed)
	}
}

func TestReconcile_repairs_missing(t *testing.T) {
	ts := newTestSetup(t)

	// Write two documents to the store without indexing.
	ts.writeStoreDoc(t, "repair-alpha", "Alpha content for repair test.")
	ts.writeStoreDoc(t, "repair-beta", "Beta content for repair test.")

	report, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("RunOnce: %v", err)
	}

	if report.MissingReindexed != 2 {
		t.Errorf("expected 2 missing reindexed, got %d", report.MissingReindexed)
	}

	// After reconciliation, the index should contain the documents.
	indexed, err := ts.index.IndexedPaths()
	if err != nil {
		t.Fatalf("IndexedPaths: %v", err)
	}
	if len(indexed) < 2 {
		t.Errorf("expected at least 2 indexed paths after repair, got %d", len(indexed))
	}
}

func TestReconcile_removes_orphans(t *testing.T) {
	ts := newTestSetup(t)

	// Index two documents, then delete both from the store.
	p1 := ts.indexDoc(t, "orphan-one", "First orphan document.")
	p2 := ts.indexDoc(t, "orphan-two", "Second orphan document.")
	if err := ts.store.Delete(context.Background(), p1); err != nil {
		t.Fatalf("delete p1: %v", err)
	}
	if err := ts.store.Delete(context.Background(), p2); err != nil {
		t.Fatalf("delete p2: %v", err)
	}

	report, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("RunOnce: %v", err)
	}

	if report.OrphanedDeleted < 2 {
		t.Errorf("expected at least 2 orphans deleted, got %d", report.OrphanedDeleted)
	}

	// After reconciliation, the index should be empty (or have no raw/documents paths).
	indexed, err := ts.index.IndexedPaths()
	if err != nil {
		t.Fatalf("IndexedPaths: %v", err)
	}
	for _, p := range indexed {
		if p == string(p1) || p == string(p2) {
			t.Errorf("orphaned path %s still present in index after reconciliation", p)
		}
	}
}

func TestReconcile_concurrent_lock(t *testing.T) {
	ts := newTestSetup(t)

	// Write a document to create some work.
	ts.writeStoreDoc(t, "concurrent-doc", "Test content for concurrency.")

	var started, finished atomic.Int32
	var wg sync.WaitGroup

	// Launch multiple goroutines that all try to run reconciliation
	// simultaneously. Only one should perform the actual work.
	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			started.Add(1)
			report, err := ts.reconciler.RunOnce(context.Background())
			if err != nil {
				t.Errorf("RunOnce error: %v", err)
				return
			}
			if report.MissingReindexed > 0 || report.DriftDetected {
				finished.Add(1)
			}
		}()
	}
	wg.Wait()

	// At most one goroutine should have done actual reconciliation
	// work. The others should have returned immediately due to the lock.
	finishedCount := finished.Load()
	if finishedCount > 1 {
		t.Errorf("expected at most 1 goroutine to perform work, got %d", finishedCount)
	}
}

func TestReconcile_idempotent(t *testing.T) {
	ts := newTestSetup(t)

	// Write a document without indexing.
	ts.writeStoreDoc(t, "idempotent-doc", "Idempotent reconciliation test content.")

	// First run: should repair the missing document.
	report1, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("first RunOnce: %v", err)
	}
	if report1.MissingReindexed != 1 {
		t.Fatalf("first run: expected 1 missing reindexed, got %d", report1.MissingReindexed)
	}

	// Second run: everything should be aligned, no drift.
	report2, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("second RunOnce: %v", err)
	}
	if report2.DriftDetected {
		t.Errorf("second run: expected no drift, but drift was detected (missing=%d, orphans=%d)",
			report2.MissingReindexed, report2.OrphanedDeleted)
	}
	if report2.MissingReindexed != 0 {
		t.Errorf("second run: expected 0 missing reindexed, got %d", report2.MissingReindexed)
	}
	if report2.OrphanedDeleted != 0 {
		t.Errorf("second run: expected 0 orphans deleted, got %d", report2.OrphanedDeleted)
	}
}

func TestReconcile_empty_store_and_index(t *testing.T) {
	ts := newTestSetup(t)

	// Both store and index are empty. Reconciliation should be a no-op.
	report, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("RunOnce: %v", err)
	}

	if report.TotalDocuments != 0 {
		t.Errorf("expected 0 total documents, got %d", report.TotalDocuments)
	}
	if report.TotalIndexed != 0 {
		t.Errorf("expected 0 total indexed, got %d", report.TotalIndexed)
	}
	if report.DriftDetected {
		t.Errorf("expected no drift on empty store/index")
	}
	if report.MissingReindexed != 0 {
		t.Errorf("expected 0 missing reindexed, got %d", report.MissingReindexed)
	}
	if report.OrphanedDeleted != 0 {
		t.Errorf("expected 0 orphans deleted, got %d", report.OrphanedDeleted)
	}
}

func TestReconcile_max_repairs_circuit_breaker(t *testing.T) {
	ts := newTestSetup(t)

	// Override max repairs to 2.
	ts.reconciler.cfg.MaxRepairs = 2

	// Write 5 documents without indexing.
	for i := 0; i < 5; i++ {
		ts.writeStoreDoc(t, fmt.Sprintf("breaker-%d", i), fmt.Sprintf("Breaker content %d.", i))
	}

	report, err := ts.reconciler.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("RunOnce: %v", err)
	}

	// Should cap at 2 repairs due to circuit breaker.
	if report.MissingReindexed > 2 {
		t.Errorf("expected at most 2 missing reindexed (circuit breaker), got %d", report.MissingReindexed)
	}
}

func TestReconcile_context_cancellation(t *testing.T) {
	ts := newTestSetup(t)

	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Millisecond)
	defer cancel()

	// Give the timeout time to fire.
	time.Sleep(5 * time.Millisecond)

	_, err := ts.reconciler.RunOnce(ctx)
	if err == nil {
		// Context may have been cancelled during listing.
		// Either an error or a zero-repair report is acceptable.
	}
}

func TestNewReconciler_requires_store(t *testing.T) {
	_, err := NewReconciler(ReconcileConfig{Index: &search.Index{}})
	if err == nil {
		t.Fatalf("expected error when Store is nil")
	}
}

func TestNewReconciler_requires_index(t *testing.T) {
	store := mem.New()
	defer func() { _ = store.Close() }()
	_, err := NewReconciler(ReconcileConfig{Store: store})
	if err == nil {
		t.Fatalf("expected error when Index is nil")
	}
}

File: sdks/ts/memory/src/ingest/reconcile.test.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Unit tests for the self-healing reconciliation module. Uses MockStore
 * and MockSearchIndex from the shared test helpers, plus a real SQLite
 * search index for integration-style assertions.
 */

import { afterEach, describe, expect, it } from 'vitest'
import { createHashEmbedder } from '../llm/hashembed.js'
import { type SearchIndex, createSearchIndex } from '../search/index.js'
import { type Path, toPath } from '../store/index.js'
import { createMemStore } from '../store/memstore.js'
import { type Reconciler, createReconciler } from './reconcile.js'
import { testLogger } from './test-helpers.js'

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const RAW_DOCUMENTS_PREFIX = 'raw/documents'

const indices: SearchIndex[] = []
const reconcilers: Reconciler[] = []

const fresh = async (dim = 32): Promise<SearchIndex> => {
  const idx = await createSearchIndex({ dbPath: ':memory:', vectorDim: dim })
  indices.push(idx)
  return idx
}

afterEach(async () => {
  for (const r of reconcilers) {
    r.stop()
  }
  reconcilers.length = 0
  while (indices.length > 0) {
    const idx = indices.pop()
    if (idx !== undefined) await idx.close()
  }
})

/**
 * Write a markdown document to the store at the canonical raw/documents
 * path, simulating a successful ingest store step.
 */
const writeStoreDoc = async (
  store: ReturnType<typeof createMemStore>,
  slug: string,
  body: string,
): Promise<string> => {
  const path = `${RAW_DOCUMENTS_PREFIX}/${slug}.md`
  const content = `---\ntitle: "${slug}"\n---\n\n${body}\n`
  await store.write(toPath(path), Buffer.from(content, 'utf8'))
  return path
}

/**
 * Write a document to the store AND index it in the search index,
 * simulating a complete ingest.
 */
const indexDoc = async (
  store: ReturnType<typeof createMemStore>,
  searchIndex: SearchIndex,
  slug: string,
  body: string,
): Promise<string> => {
  const path = await writeStoreDoc(store, slug, body)
  const embedder = createHashEmbedder({ dim: 32 })
  const texts = [body]
  const embeddings = await embedder.embed(texts)
  searchIndex.upsertChunks(
    texts.map((text, i) => ({
      id: `${slug}:${i}`,
      path,
      ordinal: i,
      title: slug,
      content: text,
      metadata: { documentId: slug, brainId: 'test' },
      ...(embeddings[i] !== undefined ? { embedding: embeddings[i] } : {}),
    })),
  )
  return path
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('createReconciler', () => {
  it('detects documents in store but missing from search index', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()
    const reindexed: string[] = []

    await writeStoreDoc(store, 'missing-doc', 'This document was stored but never indexed.')

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      reindexFn: async (path) => {
        reindexed.push(path)
      },
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.totalDocuments).toBe(1)
    expect(report.driftDetected).toBe(true)
    expect(report.missingReindexed).toBe(1)
    expect(report.orphanedDeleted).toBe(0)
    expect(reindexed).toHaveLength(1)
    expect(reindexed[0]).toContain('missing-doc')
  })

  it('detects orphaned index entries with no corresponding store document', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()

    // Index a document, then remove it from the store.
    const path = await indexDoc(store, searchIndex, 'orphan-doc', 'Content that will be orphaned.')
    await store.delete(toPath(path))

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.driftDetected).toBe(true)
    expect(report.orphanedDeleted).toBeGreaterThanOrEqual(1)
    expect(report.missingReindexed).toBe(0)

    // Verify the orphaned entry was removed.
    const remaining = searchIndex.indexedPaths()
    expect(remaining).not.toContain(path)
  })

  it('repairs missing documents by calling reindexFn', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()
    const reindexed: string[] = []

    await writeStoreDoc(store, 'repair-alpha', 'Alpha content for repair test.')
    await writeStoreDoc(store, 'repair-beta', 'Beta content for repair test.')

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      reindexFn: async (path) => {
        reindexed.push(path)
        // Simulate re-indexing by actually inserting a chunk.
        const embedder = createHashEmbedder({ dim: 32 })
        const text = (await store.read(toPath(path) as Path)).toString('utf8')
        const embeddings = await embedder.embed([text])
        searchIndex.upsertChunks([
          {
            id: `reindex:${path}:0`,
            path,
            ordinal: 0,
            title: path,
            content: text,
            metadata: { documentId: path, brainId: 'test' },
            ...(embeddings[0] !== undefined ? { embedding: embeddings[0] } : {}),
          },
        ])
      },
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.missingReindexed).toBe(2)
    expect(reindexed).toHaveLength(2)
  })

  it('removes orphaned chunks from the index', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()

    const path1 = await indexDoc(store, searchIndex, 'orphan-one', 'First orphan document.')
    const path2 = await indexDoc(store, searchIndex, 'orphan-two', 'Second orphan document.')

    // Delete both documents from the store.
    await store.delete(toPath(path1))
    await store.delete(toPath(path2))

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.orphanedDeleted).toBeGreaterThanOrEqual(2)

    const remaining = searchIndex.indexedPaths()
    expect(remaining).not.toContain(path1)
    expect(remaining).not.toContain(path2)
  })

  it('prevents concurrent reconciliation via brain-level lock', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()
    let reindexCount = 0

    await writeStoreDoc(store, 'concurrent-doc', 'Content for concurrency test.')

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      reindexFn: async () => {
        // Simulate slow re-indexing.
        await new Promise((resolve) => setTimeout(resolve, 50))
        reindexCount++
      },
    })
    reconcilers.push(reconciler)

    // Launch multiple reconciliation runs in parallel.
    const results = await Promise.all([
      reconciler.runOnce(),
      reconciler.runOnce(),
      reconciler.runOnce(),
    ])

    // At most one should have performed actual work.
    const workingRuns = results.filter((r) => r.driftDetected)
    expect(workingRuns.length).toBeLessThanOrEqual(1)
  })

  it('is idempotent: running twice produces the same result', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()

    await writeStoreDoc(store, 'idempotent-doc', 'Idempotent reconciliation test content.')

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      reindexFn: async (path) => {
        // Simulate re-indexing by inserting a chunk.
        const text = (await store.read(toPath(path) as Path)).toString('utf8')
        const embedder = createHashEmbedder({ dim: 32 })
        const embeddings = await embedder.embed([text])
        searchIndex.upsertChunks([
          {
            id: `idempotent:${path}:0`,
            path,
            ordinal: 0,
            title: path,
            content: text,
            metadata: { documentId: path, brainId: 'test' },
            ...(embeddings[0] !== undefined ? { embedding: embeddings[0] } : {}),
          },
        ])
      },
    })
    reconcilers.push(reconciler)

    // First run: should repair the missing document.
    const report1 = await reconciler.runOnce()
    expect(report1.missingReindexed).toBe(1)
    expect(report1.driftDetected).toBe(true)

    // Second run: should find everything aligned.
    const report2 = await reconciler.runOnce()
    expect(report2.driftDetected).toBe(false)
    expect(report2.missingReindexed).toBe(0)
    expect(report2.orphanedDeleted).toBe(0)
  })

  it('reports no drift when store and index are both empty', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.totalDocuments).toBe(0)
    expect(report.totalIndexed).toBe(0)
    expect(report.driftDetected).toBe(false)
    expect(report.missingReindexed).toBe(0)
    expect(report.orphanedDeleted).toBe(0)
    expect(report.errors).toBe(0)
  })

  it('respects maxRepairs circuit breaker', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()
    let reindexCount = 0

    // Write 5 documents without indexing.
    for (let i = 0; i < 5; i++) {
      await writeStoreDoc(store, `breaker-${i}`, `Breaker content ${i}.`)
    }

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      maxRepairs: 2,
      reindexFn: async () => {
        reindexCount++
      },
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    // Should cap at 2 repairs.
    expect(report.missingReindexed).toBeLessThanOrEqual(2)
    expect(reindexCount).toBeLessThanOrEqual(2)
  })

  it('counts errors when reindexFn throws', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()

    await writeStoreDoc(store, 'error-doc', 'Document that will fail re-indexing.')

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      reindexFn: async () => {
        throw new Error('simulated re-index failure')
      },
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.driftDetected).toBe(true)
    expect(report.errors).toBe(1)
    expect(report.missingReindexed).toBe(0)
  })

  it('reports accurate drift metrics for mixed missing and orphaned', async () => {
    const store = createMemStore()
    const searchIndex = await fresh()
    const reindexed: string[] = []

    // Create a document in store but not indexed (missing).
    await writeStoreDoc(store, 'mixed-missing', 'Missing from index.')

    // Create a document in both store and index, then delete from store (orphaned).
    const orphanPath = await indexDoc(
      store,
      searchIndex,
      'mixed-orphan',
      'Will become orphaned.',
    )
    await store.delete(toPath(orphanPath))

    const reconciler = createReconciler({
      store,
      searchIndex,
      logger: testLogger,
      reindexFn: async (path) => {
        reindexed.push(path)
      },
    })
    reconcilers.push(reconciler)

    const report = await reconciler.runOnce()

    expect(report.driftDetected).toBe(true)
    expect(report.missingReindexed).toBe(1)
    expect(report.orphanedDeleted).toBeGreaterThanOrEqual(1)
    expect(report.totalDocuments).toBe(1)
    expect(reindexed).toHaveLength(1)
  })
})
